rfcs/framework-0072-pluggable-transfers.md

```diff
@@ -0,0 +1,507 @@
+the `Transfer` class).
+
+### Transfer types
+
+We will convert `TransferType` from an enum into a regular class.
+Will have string constants for built-in types, e.g., the `TransferType.LOCAL` will be replaced with `TRANSFER_TYPE_LOCAL`.
+
+### Transfer implementation
+
+A new `Transfer` base class will be introduced with the following API:
+
+```python
+class Transfer:
+    transfer_type: str
+    Schema: ma.Schema
+
+    def __init__(self, record: Record, key: str, file_service: FileService, uow=None): ...
+    # file upload operations
+    def init_file(self, record, file_metadata): ...
+    def set_file_content(self, stream, content_length): ...
+    def commit_file(self): ...
+    def delete_file(self): ...
+
+    # upload status
+    def status(self): ...
+
+    # send file to client
+    def send_file(self, *, restricted, as_attachment): ...
+
+    # get extra links that the transfer may provide
+    def expand_links(self, identity, self_url): ...
+```
+
+Then for each built-in transfer, a dedicated implementation will be created.
+
+### TransferRegistry
+
+A `TransferRegistry` class will be introduced, modeled after the
+`ServiceRegistry`. A new proxy with pre-filled transfers will be provided as
+`current_transfer_registry`.
+
+The transfer registry will be configured through the following options:
+
+```python
+RECORDS_RESOURCES_TRANSFERS = [
+    "invenio_records_resources.services.files.transfer.LocalTransfer",
+    "invenio_records_resources.services.files.transfer.FetchTransfer",
+    "invenio_records_resources.services.files.transfer.RemoteTransfer",
+    "invenio_records_resources.services.files.transfer.MultipartTransfer",
+]
+"""List of transfer classes to register."""
+
+
+RECORDS_RESOURCES_DEFAULT_TRANSFER_TYPE = "L"
+"""Default transfer class to use. 
+One of 'L' (local), 'F' (fetch), 'R' (point to remote), 'M' (multipart)."""
+```
+
+### Initial vs. File schema
+
+The distinction between `InitialFileSchema` and `FileSchema` has been blurred by
+recent changes in **invenio-records-resources**, where `FileSchema` is now used
+for `init_files` as well.
+
+To clarify, we will turn `InitialFileSchema` into a mixin that is applied to
+`FileSchema` automatically when a new file is created.
+
+This approach lets us mark all “read only after creation” fields, such as
+`key`, `storage_class`, `mimetype`, `size`, and `checksum`, as `dump_only` in
+`FileSchema`. In the `InitialFileSchema`, we will re-enable these fields for
+writing by marking them as `load_only`.
+
+This ensures that users can provide their own file schemas for different use
+cases (e.g., media files vs. regular files) while still preserving the correct
+handling of initialization parameters.
+
+We will also introduce a `TransferSchema` to handle the new `transfer` field.
+It will allow using different schemas for each transfer type. To achieve this,
+we will leverage `marshmallow-oneofschema`, which is already included in the
```

Comment:
I personally find the usage of  `oneofschema` a bit complex, so I would like to suggest documenting it well in the code.